diesel_derives/src/associations.rs

```diff
@@ -2,55 +2,84 @@ use syn;
+                    .map(|options| expand_belongs_to(&model, options))
+                    .unwrap_or_else(|e| e),
+            )
+        } else {
+            None
         }
-    }
+    });
 
-    quote!(#(#derived_associations)*)
+    wrap_item_in_const(
+        model.dummy_const_name("ASSOCIATIONS"),
+        quote!(#(#derived_associations)*),
+    )
 }
 
 fn expand_belongs_to(model: &Model, options: AssociationOptions) -> quote::Tokens {
     let parent_struct = options.name;
     let struct_name = &model.name;
+    let fields = model.attrs.as_slice();
 
     let foreign_key_name = options
         .foreign_key_name
         .unwrap_or_else(|| to_foreign_key(parent_struct.as_ref()));
+    let foreign_key_name = &foreign_key_name;
     let child_table_name = model.table_name();
-    let fields = model.attrs.as_slice();
+    let child_table_name = &child_table_name;
+    let foreign_key_attr = fields
+        .iter()
+        .find(|attr| attr.column_name().as_ref() == foreign_key_name.as_ref())
+        .unwrap();
+    let foreign_key_ty = &foreign_key_attr.ty;
+
+    // we need to special case foreign keys on with an Option type
+    // to allow self referencing joins
+    let (foreign_key, foreign_key_ty) = match *foreign_key_ty {
+        syn::Ty::Path(None, ref p) if p.segments[0].ident == "Option" => {
+            let segment = &p.segments[0];
+            let t = match segment.parameters {
+                syn::PathParameters::AngleBracketed(ref p) => p.types[0].clone(),
+                syn::PathParameters::Parenthesized(_) => unreachable!(),
+            };
+            (quote!(self.#foreign_key_name.as_ref()), t)
+        }
+        ref t => (quote!(Some(&self.#foreign_key_name)), t.clone()),
+    };
 
-    quote!(BelongsTo! {
-        (
-            struct_name = #struct_name,
-            parent_struct = #parent_struct,
-            foreign_key_name = #foreign_key_name,
-            child_table_name = #child_table_name,
-        ),
-        fields = [#(#fields)*],
-    })
+    quote!(
+        impl diesel::associations::BelongsTo<#parent_struct> for #struct_name {
+            type ForeignKey = #foreign_key_ty;
+            type ForeignKeyColumn = #child_table_name::#foreign_key_name;
+
+            fn foreign_key(&self) -> Option<&Self::ForeignKey> {
+                #foreign_key
+            }
+
+            fn foreign_key_column() -> Self::ForeignKeyColumn {
+                #child_table_name::#foreign_key_name
+            }
+        }
+    )
 }
 
 struct AssociationOptions {
     name: syn::Ident,
     foreign_key_name: Option<syn::Ident>,
 }
 
-fn build_association_options(
-    attr: &syn::Attribute,
-    association_kind: &str,
-) -> Result<AssociationOptions, String> {
-    let usage_error = Err(format!(
-        "`#[{}]` must be in the form `#[{}(table_name, option=value)]`",
-        association_kind, association_kind
+fn build_association_options(attr: &syn::Attribute) -> Result<AssociationOptions, quote::Tokens> {
```

Comment:
The changes to this function feel unrelated to the rest of this PR. Can we leave them out?

Reply:
We can move them to a separate pull request. These changes are unrealated to the actual porting, but include
* some cleanup (removing a unused function parameter)
* improved error handling, compile error and checking the syntax of the `#[belongs_to()]` attribute more deeply

So just open a follow up pullrequest containing this changes, as soon as we merged this one, or only put those changes into a separate commit?